packages/core/admin/utils/get-custom-app-config-file.js

```diff
@@ -0,0 +1,23 @@
+'use strict';
+
+const { join } = require('path');
+const fse = require('fs-extra');
+const { isUsingTypeScript } = require('@strapi/typescript-utils');
+
+/**
+ * Retrieve the custom admin entry file name
+ * @param {String} dir - Directory of the admin panel
+ * @returns String
+ */
+const getCustomAppConfigFile = async dir => {
+  const adminSrcPath = join(dir, 'src', 'admin');
+  const useTypeScript = await isUsingTypeScript(adminSrcPath, 'tsconfig.json');
+
+  const files = await fse.readdir(adminSrcPath);
+
+  const appRegex = new RegExp(`app.${useTypeScript ? 't' : 'j'}sx?$`);
+
+  return files.find(file => file.match(appRegex));
```

Comment:
a .js file is still valid for typescript in theory so I wouldn't prevent a user to have .js files

Reply:
In theory I would agree, but if you're using an app.tsx and you compile in place, you'll end up with both app.jsx and app.tsx. How do you choose which one is the real entry point that needs to be interpreted by TS?